### ISSUE TITLE
Optuna integration significantly reduces performance and limits number of parallel trials in Ray Tune

### ISSUE THREAD

**Issue by Aricept094** [optuna/optuna#5611]
### Expected behavior

Optuna integration should maintain performance comparable to other search algorithms, utilize CPU resources efficiently, and allow for multiple trials as specified.

### Environment

Ray version: 2.32.0
Python version: 3.11.8
Operating System: Linux-6.6.36.3-microsoft-standard-WSL2-x86_64-with-glibc2.35
Optuna version: 3.6.1

Detailed OS Information:
OS: Linux
OS Version: https://github.com/ray-project/ray/pull/1 SMP PREEMPT_DYNAMIC Sat Jun 29 07:01:04 UTC 2024
OS Release: 6.6.36.3-microsoft-standard-WSL2
Machine: x86_64
Processor: x86_64

### Error messages, stack traces, or logs

```shell
No specific error.
```

### Steps to reproduce

The execution time for 100 trials shows a dramatic difference:

With Optuna: 41 seconds
Without Optuna: 9 seconds

```python
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import ray
from ray import tune
from ray.tune.search.optuna import OptunaSearch
from ray import train, tune

def generate_data(n_samples=1000):
    X = np.random.rand(n_samples, 5)
    y = 2 * X[:, 0] + 3 * X[:, 1] - X[:, 2] + 0.5 * X[:, 3] - 1.5 * X[:, 4] + np.random.normal(0, 0.1, n_samples)
    return X, y

def train_random_forest(config):
    X, y = generate_data()
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)
    
    rf = RandomForestRegressor(
        n_estimators=config["n_estimators"],
    )
    
    rf.fit(X_train, y_train)
    y_pred = rf.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    
    train.report({"mean_squared_error": mse,})


def main():
    ray.init()
    
    config = {
        "n_estimators": tune.randint(10, 200),
    }
        
    analysis = tune.run(
        train_random_forest,
        config=config,
        num_samples=100, 
        search_alg=OptunaSearch(),
        metric="mean_squared_error",
        mode="min",
        reuse_actors=True
    )
    
    print("Best config:", analysis.best_config)
    print("Best MSE:", analysis.best_result["mean_squared_error"])

if __name__ == "__main__":
    main()

```

### Additional context (optional)

Also referenced here in ray

[](https://github.com/ray-project/ray/issues/46965)

**Comment by Aricept094** [comment]
[](https://github.com/ray-project/ray/issues/46965)

**Comment by Aricept094** [comment]
[https://github.com/ray-project/ray/issues/46965](url)

**Comment by Aricept094** [comment]
[](https://github.com/ray-project/ray/issues/46965#issue-2447221959)

**Comment by nabenabe0928** [comment]
Hi, thank you for your report!
Could you tell me what you meant by `without Optuna` and how to run the code above without Optuna?

**Comment by Aricept094** [comment]
> Hi, thank you for your report!
> 
> Could you tell me what you meant by `without Optuna` and how to run the code above without Optuna?

Hey , thanks for fast response. 

Sorry for being a bit vague. 

"With optuna" I meant using OptunaSearch() as the search algorithm in the "def main" and without optuna I meant just removing the line so ray defaults to its base search algorithm. 

Just copy and pasting the code should work .

**Comment by nabenabe0928** [comment]
@Aricept094 Thank you for the note! I will check the cause and get back to you:)

**Comment by kAIto47802** [comment]
Optuna DOES support both multi-processing and multi-threading, but Optuna study must be specified with an RDB or journal storage to make use of it.
I checked your code, but Ray uses multi-processing while Ray with Optuna did not and Optuna itself was not the cause of the slowdown. 
Ray itself defines the storage as in-memory storage [here](https://github.com/ray-project/ray/blob/3109a7eae635e8567a25a9fafa9645d4d7f8237b/python/ray/tune/search/optuna/optuna_search.py#L383), meaning that we cannot do anything about it without letting the Ray developers to handle this issue.
With that being said, it is understandable that you would like to use distributed optimization, so I would suggest two alternative options:

1. Use Optuna without Ray, please check [the tutorial](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/004_distributed.html)
2. Raise an issue to the ray repository.

I do not exactly know your use case, but if you do not need to use Ray, it would be more stable to use Optuna as is.

**Comment by Aricept094** [comment]
> @Aricept094
> 
> Optuna DOES support both multi-processing and multi-threading, but Optuna study must be specified with an RDB or journal storage to make use of it.
> 
> I checked your code, but Ray uses multi-processing while Ray with Optuna did not and Optuna itself was not the cause of the slowdown. 
> 
> Ray itself defines the storage as in-memory storage [here](https://github.com/ray-project/ray/blob/3109a7eae635e8567a25a9fafa9645d4d7f8237b/python/ray/tune/search/optuna/optuna_search.py#L383), meaning that we cannot do anything about it without letting the Ray developers to handle this issue.
> 
> With that being said, it is understandable that you would like to use distributed optimization, so I would suggest two alternative options:
> 
> 
> 
> 1. Use Optuna without Ray, please check [the tutorial](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/004_distributed.html)
> 
> 2. Raise an issue to the ray repository.
> 
> 
> 
> I do not exactly know your use case, but if you do not need to use Ray, it would be more stable to use Optuna as is. 
> 
> 

Thank you very much for detailed response. i have raised the issue on ray repo but it didn't get much attention, i will try other methods too .

**Comment by nabenabe0928** [comment]
@Aricept094 
I personally would recommend using Optuna as is if your usecase does not require Ray strictly.
We are always welcome to hear your opinion if Optuna by itself does not support your usecase!

**Comment by nabenabe0928** [comment]
@Aricept094 
Could you please try the code above and tell me if it works?
If it works, I think it means this issue is resolved.

**Comment by nabenabe0928** [comment]
> [!NOTE]
> Btw, please note that it is not a good practice to tune `n_estimators` of random forest as it controls how much you would like to train the model, meaning that usually the performance increases monotonically as `n_estimators` increases.

**Comment by nabenabe0928** [comment]
I would like @Aricept094 to check [here](https://optuna.readthedocs.io/en/latest/tutorial/10_key_features/002_configurations.html#defining-parameter-spaces), but it is very straightforward to implement conditional search space in Optuna.

**Comment by Aricept094** [comment]
> I would like @Aricept094 to check [here](https://optuna.readthedocs.io/en/latest/tutorial/10_key_features/002_configurations.html#defining-parameter-spaces), but it is very straightforward to implement conditional search space in Optuna.
> 
> 
> 
> ```python
> 
> def objective(trial):
> 
>     used_algorithm = trial.suggest_categorical("algorithm", choices=["svm", "logistic"])
> 
>     if used_algorithm == "svm":
> 
>         C = trial.suggest_float(...)
> 
>         kernel = trial.suggest_categorical(...)
> 
>         degree = trial.suggest_int(...)
> 
>         gamma = trial.suggest_categorical(...)
> 
>         coef0 = trial.suggest_float(...)
> 
>         shrinking = trial.suggest_categorical(...)
> 
>         tol = trial.suggest_float(...)
> 
>         decision_function_shape = trial.suggest_categorical(...)
> 
>         # Instantiate SVM here.
> 
>     elif used_algorithm == "logistic":
> 
>         C = trial.suggest_float(...)
> 
>         tol = trial.suggest_float(...)
> 
>         solver = trial.suggest_categorical(...)
> 
>         fit_intercept = trial.suggest_categorical(...)
> 
>         intercept_scaling = trial.suggest_float(...)
> 
> 
> 
>         if logistic_solver == "liblinear":
> 
>             penalty = trial.suggest_categorical("liblinear_penalty", choices=["l1", "l2"])
> 
>         elif logistic_solver == "saga":
> 
>             penalty = trial.suggest_categorical("saga_penalty", choices=["l2", None])
> 
>         else:
> 
>             penalty = trial.suggest_categorical("penalty", ["l1", "l2", "elasticnet", None])
> 
> 
> 
>         if logistic_solver == "liblinear" and penalty == "l2":
> 
>             logistic_dual = trial.suggest_categorical(...)
> 
>         else:
> 
>             logistic_dual = False
> 
>         
> 
>         if penalty == "elasticnet":
> 
>             l1_ratio = trial.suggest_float(...)
> 
>         else:
> 
>             l1_ratio = None
> 
>         
> 
>         # Instantiate Logistic Model here.
> 
> 
> 
>     ...
> 
>     return error_rate
> 
> 
> 
> ```

- [ ] Oh great, I've got nothing to say . Didn't expect it to be that simple . Thanks for your help , i will try to convert my codes to solely use optuna and if i find a feature that is not available in optuna i will let you guys know.

**Comment by nabenabe0928** [comment]
Great:)
It seems like the issue is resolved, so I will close this issue!

**Comment by Aricept094** [comment]
The performance issues described at #5611 should be resolved by ray-project/ray#48547, which adds support for configuring custom Optuna storage in OptunaSearch. This allows using RDB or other storage backends that support proper parallelization instead of being limited to in-memory storage.


This change was merged on Dec 10, 2024 and should be available in the next Ray release.

You will be able to configure storage like this:


```python
from ray.tune.search.optuna import OptunaSearch
import optuna

storage = optuna.storages.RDBStorage(
    "sqlite:///example.db",
    heartbeat_interval=1
)

search_alg = OptunaSearch(
    storage=storage,
    study_name="example_study"
)`

**Comment by Aricept094** [comment]
I’d like to update the discussion regarding the performance and parallelization limitations when using Optuna with Ray Tune. As noted in my earlier report [#5611](https://github.com/ray-project/ray/issues/5611), using the default in‑memory storage for Optuna in Ray Tune resulted in significant performance degradation and limited parallel trial execution. In that thread, it was pointed out that for Optuna to leverage multi‑processing and multi‑threading, the study must be configured with an external storage backend (such as RDB or Journal storage).

Recently, [ray-project/ray#48547](https://github.com/ray-project/ray/pull/48547) was merged (Dec 10, 2024) to add support for configuring custom Optuna storage via the storage parameter in OptunaSearch. According to the update, you can now configure an external storage backend like this:

```python
from ray.tune.search.optuna import OptunaSearch
import optuna

storage = optuna.storages.RDBStorage(
    "sqlite:///example.db",
    heartbeat_interval=1
)

search_alg = OptunaSearch(
    storage=storage,
    study_name="example_study"
)

```


The Problem at Hand

Despite this improvement, in my experiments the trials still execute sequentially rather than in parallel. For example, when launching an experiment with a large number of trials (e.g., 100–200 trials expected to run concurrently) the dashboard shows only one trial in a “PENDING” state at a time. This is contrary to the behavior observed when using other search algorithms in Ray Tune, where multiple trials are scheduled simultaneously (leveraging available CPU resources).

I’ve tried configuring both the external storage (as suggested) and Tune’s own parallelism controls (such as setting max_concurrent_trials in the TuneConfig and even calling set_max_concurrency() on the Optuna searcher), yet the observed behavior remains unchanged: the trials continue to run sequentially.

In previous discussions (see [#5611](https://github.com/ray-project/ray/issues/5611)), the performance issues were attributed to Ray’s in‑memory storage for Optuna. The new update should, in theory, allow for proper multi‑processing by using external storage. However, in practice, even with the storage configured as above, the trials remain queued and are executed one after the other.

I want to thank the devs (@kAIto47802 and @nabenabe0928 ) for their kindness and help in teaching me how to run Optuna in parallel—by launching multiple Python script instances, I was able to parallelize hyperparameter tuning successfully using Optuna alone. However, I’ve now encountered a new challenge with tracking these runs through Weights & Biases (wandb).

Even when using the wandb integration library, each new instance of the Python script registers as a separate wandb run. For example, if I have 8 CPU cores (and thus run 8 instances concurrently), wandb creates 8 separate runs. This behavior prevents me from tracking the overall progress of my hyperparameter tuning in a single consolidated wandb run. Instead of seeing one run that aggregates all trial data, I end up with multiple, fragmented runs, which makes it difficult to get a holistic view of the tuning process so i was reverted back to using Ray but with out optuna so i can have parallelization with wandb .
